### ISSUE TITLE
Decide how to compute exact Hessians

### ISSUE THREAD

**Issue by jeff-regier** [jeff-regier/Celeste.jl#7]
We should decide how to compute exact Hessians. Hessians may be computed either 1) manually, 2) automatically, by JuMP/ReverseDiffSparse, or 3) automatically, by forking JuMP and modifying it to parse a BUGS-like specification of our model. To decide, it'd helpful to understand whether our objective function can be written in the current JuMP modeling language.

**Comment by rgiordan** [comment]
I think as a first step it might be worth trying BFGS preconditioning with only the diagonal of the true Hessian to get the problem "scale" right.  That will be easy to calculate by hand.

**Comment by jeff-regier** [comment]
Great idea!

**Comment by jeff-regier** [comment]
On second thought, if you manually compute the diagonal of the Hessian, convert a bounded optimization problem to an unbounded one, and call BFGS with preconditioning, you've actually done a good chunk of the work needed to manually compute the exact Hessian and use Newton's method. If it's only going to take a week or so to get diagonally preconditioned BFGS running, you might as well just go for it. But if it's a bigger undertaking, you might want to spend like a day looking at what's possible with JuMP first. Because if JuMP can work for us and we use it, we'd presumably end up with really concise code, that would be easy to modify in the future...it takes the derivatives, it finds the Hessian, and then it calls a really mature optimization library (IpOpt) that naively support all the kinds of constraints we might want. We'd be able to delete `OptimizeElbo.jl` almost entirely, `ElboDeriv.jl` might become half its current length, and we wouldn't need a lot of the stuff in `CelesteTypes.jl` either.

**Comment by rgiordan** [comment]
I think right now it might take me a week just because I'm still getting
used to the basics of Julia, which won't be mitigated by looking into JuMP.

**Comment by jeff-regier** [comment]
That's fine, it sounds to me like more than a week of work, even if you were familiar with Julia. My point was just that JuMP + IpOpt and manually computed second derivatives + BFGS are alternatives...makes sense to at least consider whether JuMP could work before spending a lot of time on a competing (and likely tedious) approach. I could spend some time later this week or next week evaluating whether JuMP might work for our problem.

**Comment by rgiordan** [comment]
Learning about JuMP sounds like more fun than computing second derivatives,
so if you think it's best to start there I'll check it out before doing any
calculations of my own.

**Comment by rgiordan** [comment]
I spent some time learning about JuMP (and how to debug Julia package install errors) today.  I am tentatively optimistic that it would be possible to do Celeste in JuMP, but based on the basic documentation at least I suspect it would not be a quick re-write of the existing code.  But maybe I've missed something or you can think of some meta-programming tricks.

The problem as I see it is how to keep the structure of ElboDeriv.jl but have everything assign JuMP expression objects rather than return values.  As far as I can tell, you can't do something like

```
function MyFun(x)
   x + 5
end
@defNLExpr(x_plus_5, MyFun(x))
```

However, you can chain together the result of defNLExpr.  So you could do something like

```
function AddFive(i)
  @defNLExpr(x_plus_5[i], x[i] + 5)
end

function Square(i)
  @defNLExpr(x_plus_5_squared[i], x_plus_5[i]^2
end
```

It would be nice to do that without having to index everything, i.e. to have something like the imaginary

```
function AddFiveAndStore(x, result_name)
  @def(result_name, x + 5)
end
```

Maybe there's some metaprogramming thing that would let you do that.  Though as Miles mentioned, Julia doesn't handle flat sums very well, so that might not be advisable anyway.

Finally, even though this seems possible in principle, I couldn't really get even pretty simple things like this to actually work.  (This is partly, I suspect, because NLOpt doesn't play very nicely with JuMP and I have been unable to get Ipopt to work on my mac due to persistent fortran compiler errors.)  For example, the very simple optimal control example from the website does not run with NLOpt, and the error messages are pretty cryptic.

So in short, I think JuMP is doable, but not faster than doing the Hessian diagonal by hand.  How about I try some basic stuff on linux when I go to Berkeley tomorrow, and if the above ideas are actually possible I email Miles, but in the meantime focus on trying to get unconstrained pre-conditioned BFGS working so you can make quicker progress on the galaxy model?

**Comment by jeff-regier** [comment]
I spent a bit of time looking at JuMP too...don't want to step on your toes here....just wanted to make sure we're going the right direction while you're getting started. Check out this toy program:

```
using JuMP
using Ipopt

const N = 5000000
const A = randn(N); 

m = Model(solver=IpoptSolver())
@defVar(m, -10 <= x <= 10)
@setNLObjective(m, Min, sum{exp(x*A[i]), i=1:N})   #This is like the sum over pixels in our objective.
solve(m)
```

It's amazing...like 5--15 iterations until it converges, all in less than 1 minute on a very old laptop. The toy problem above is simple, but it's not unlike our actual objective function: a sum over ~5 million pixels, expensive calls to `exp`. I also tried  calling Ipopt directly, with manually computed gradients/hessian for this toy problem---it's no faster.

If you exclude all the derivatives and comments from `ElboDeriv.jl` our objective is only like 100-200 lines of code. Should be a pretty straightforward mapping from it to JuMP syntax (fingers crossed), making heavy use of JuMP's `@defExpr` macro. Do you want to try working on that tomorrow?

**Comment by jeff-regier** [comment]
That Fortran compiler issue was a nightmare for me too. On linux (and probably on mac too), it's solved by deleting `lib/julia/libgfortran*` from your julia directory. Compiling Julia from source, rather than downloading the binaries also solves the problem.

**Comment by rgiordan** [comment]
Hm, I just tried deleting libgfortran.dylib and that didn't fix the
problem.  (I'm pretty sure it's using the library in /usr/lib anyway
because one of the earlier problems was it being unable to find that
library.)

If your'e curious, I opened an issue
https://github.com/JuliaOpt/Ipopt.jl/issues/30#issuecomment-76127699 with
more information.

**Comment by rgiordan** [comment]
Ok, cool.  I like the idea of JuMP if we can get it to work.  I'm hamstrung
until I can get Ipopt installed (though I'm optimistic that my linux laptop
will be easier).

I made a JuMP branch on the repo with some screwing around.  Maybe we can
both use that branch.  I'll be more convinced if we can do something like
least squares in the following way.

@defVar(m, beta[1:k])

function SetResidual(i)
  @defExpr(eps[i], y[i] - dot(beta, vec(x[i, :]))
end

for i = 1:n
  SetResidual(i)
end
@setNLObjective(m, Min, sum{eps^2, i=1:n})
solve(m)

 (This is nothing you'd ever actually do, of course, it's just a kind of
proof of concept.)

**Comment by rgiordan** [comment]
With Ipopt in hand, I got some working examples.  I'd like you to take a look, though.  Consider the code below, which doesn't work.  The error is shown -- eps_squared is not actually being defined inside the function.  Note that if this had been done in the main body of the code, i.e. not inside the function, this would have worked fine.  (See other examples in the haphazardly organized JuMPExample.jl.)

Some scoping thing?  Any thoughts?

```
using JuMP
n = 100
k = 3
x = randn(n, k)
true_beta = randn(k)
y = x * true_beta + randn(n)

function CalculateEpsilon(index)
  @defNLExpr(eps_squared[index], (y[index] - sum{x[index, j] * beta[j], j=1:k})^2)
end
m = Model()
@defVar(m, -10 <= beta[1:k] <= 10)
for i = [n]
  CalculateEpsilon(i)
end
@setNLObjective(m, Min, sum{eps_squared[i], i=1:n})
# ERROR: eps_squared not defined
status = solve(m)
```

**Comment by rgiordan** [comment]
One more concern is whether JuMP objects make a deep copy of the data that is used to construct them.  Consider this, for example:

```
m = Model()
big_data = randn(100)
@defVar(m, big_parameter)
@defNLExpr(big_transform[i=1:100], big_parameter * big_data[i]^2)
big_data = -5
print(big_transform)
print(big_data)
```

You can see that big_transform still contains all the data from big_data.  If you were to define another expression that used big_transform, maybe it, too, would contain a copy of big_data.  My Julia chops are not strong enough to check this directly, though, and maybe I'm missing something.  Any ideas for how to check?

**Comment by jeff-regier** [comment]
Unlike `@defVar`, etc., `@defNLExp` doesn't take model `m` as it's first argument. So if you put a call to `@defNLExpr` in a for loop, it won't mutate anything outside of the body of the for loop.

Here's working code:

```
using JuMP

n = 100
k = 3
x = randn(n, k)
true_beta = randn(k)
y = x * true_beta + randn(n)

m = Model()
@defVar(m, -10 <= beta[1:k] <= 10)
@defNLExpr(eps_squared[index=1:n], (y[index] - sum{x[index, j] * beta[j], j=1:k})^2)
@setNLObjective(m, Min, sum{eps_squared[i], i=1:n})
status = solve(m)
```

Everything else worked once I fixed that. Nicely done!

**Comment by jeff-regier** [comment]
I was concerned that JuMP wouldn't do things in efficient manner too, but I'm much less so now that I've run that toy example with 5,000,000 pixels that I posted yesterday. It finished as fast as my hand coded derivatives, on a laptop with only 2 GB of RAM. 2x or even 10x too much memory isn't so bad....it's the O(n) memory inefficiencies, where n is the number of pixels, that I worry most about. (For speed on the other hand 2x overhead is ok, but 10x isn't.) As long as we can process like 10-megapixels worth of images in at most few cpu minutes, I think we're in business!

**Comment by rgiordan** [comment]
So if you cant define an NLExpr in a function, do you think the ELBO can be
cleanly written as a sequence of sums without using any functions?

-- Ryan.  Sent from a phone.  Nevertheless, all typos are deliberate.

**Comment by jeff-regier** [comment]
I think the code above is cleaner without the one line function, because it only gets called once. I'm not sure about the ELBO, but less nesting could be better. A lot of quantities are just sums of something else: the elbo is the sum a likelihood and a kl. The likelihood is a sum over 5 band-specific likelihood. Each band specific likelihood is a sum over tiles. Each tile is a sum over pixels. All that might work pretty well as a flat list of commands. You might try working backwards. You can also use constraints when necessary rather than expressions. One constraint per tile even should be ok, but one constraint per pixel would likely be a problem.

**Comment by rgiordan** [comment]
So _if_ it's making a deep copy (and it very well may not be) it seems to
me that the memory problem is on the order of number of pixels \* amount of
prior data \* # of intermediate expressions, and that would be a problem.

For example, each pixel's color ELBO term is a function of the pre-computed
color priors.  If you put that ELBO term in an intermediate expression, and
JuMP stores that expression with a deep copy, then instead of one integer
per pixel (the raw poisson counts) you now have a JuMP expression
containing a copy of the whole color prior dataset, containing many floats,
for each pixel.  The same is true for the PSF prior and galaxy prior.

A test for this would be to take your 5M pixel problem and generate a bunch
of intermediate expressions that are functions of the raw pixel sizes and
see what happens.  For example, do the min of something like

@setNLObjective(m, Min, sum{alpha_x[i] + beta_x[i] + gamma_x[i], i=1:n})

...where each inner term is a JuMP NLExpr, like
@defNLExpr(alpha_x[i=1:5000000], alpha \* x[i])
@defNLExpr(beta_x[i=1:5000000], beta^2 \* x[i]^2)
@defNLExpr(gamma_x[i=1:5000000], gamma^3 \* x[i]^3)

If that uses the same amount of RAM or only marginally more than the first
version, I'll be convinced.

**Comment by rgiordan** [comment]
> I think the code above is cleaner without the one line function, because
> it only gets called once.
> 
> Sure, this was just an example to see if you could use functions at all in
> this way.  If you could, then we could maybe move to JuMP with something
> that is much like the code as-is.

Without functions the ELBO in JuMP would look much different than it does
now.  Of course, you're much more familiar with it than I am so your
intuition about what's easy and what's not is surely better than mine.

**Comment by rgiordan** [comment]
Good news, it looks like JuMP is not being as wasteful as I was worried
about.  The amount of memory allocated by JuMP (above and beyond the
original dataset) increases linearly with the amount of data, but adding
extra @NLExpr definitions doesn't increase that memory usage.

The graph shows two models optimizing sum(exp(alpha \* x[i])), one of which
has two NLExpr definitions (one for alpha \* x[i] and one for exp(of that)),
and one with only one  (the whole thing in one expression).  The x-axis is
the log10 of the data size and the y the log10 of the memory JuMP uses.
The two models are pretty much the same.  (NB I couldn't figure out how to
plot multiple colors yet.)   Code is in the JuMP branch.

So that's promising!

**Comment by rgiordan** [comment]
I wrote and checked in a JuMP implementation of a toy version of the model (1-d, point galaxies, one band, single gaussian psf).  I do think it looks promising.

I came across two issues.  One is the lgamma() function doesn't seem to be supported or at least have an error in it.  The other is stranger, and I have posted an issue about it:
https://github.com/JuliaOpt/JuMP.jl/issues/406

**Comment by jeff-regier** [comment]
This is great progress! And I'm glad the issue you posted about got fixed so quickly...the JuMP maintainers have been really helpful.

I think I see how to get the `lgamma` function working. JuMP loads differentiation rules from `Calculus.jl`. `Calculus.jl` lists `digamma(x)` as the derivative of `lgamma(x)`, but it doesn't list a derivative for `digamma(x)`, so JuMP doesn't know that `polygamma(1, x)` is the second derivative of `lgamma(x)`. The following one-line change to `Calculus/src/differentiate.jl` remedies this:

```
diff --git a/src/differentiate.jl b/src/differentiate.jl
index 49c9270..4fe3abe 100644
--- a/src/differentiate.jl
+++ b/src/differentiate.jl
@@ -166,6 +166,7 @@ symbolic_derivative_1arg_list = [
     ( :erfi,        :(  2 * exp(square(x)) / sqrt(pi)           ))
     ( :gamma,       :(  digamma(x) * gamma(x)                   ))
     ( :lgamma,      :(  digamma(x)                              ))
+    ( :digamma,     :(  polygamma(1, x)                         ))
     ( :airy,        :(  airyprime(x)                            ))  # note: only covers the 1-arg version
     ( :airyprime,   :(  airy(2, x)                              ))
     ( :airyai,      :(  airyaiprime(x)                          ))
```

**Comment by rgiordan** [comment]
Oh, awesome.  Do you mind if I post your response to the issue in JuMP.jl
https://github.com/JuliaOpt/JuMP.jl/issues/407?

I guess this should be a pull request to Calculus.jl.  Do you want to do
that or let the JuMP guys do it?  Or should I give it a shot?  (I may have
to tap you if there are questions but I'd be happy to give it a go.)

**Comment by rgiordan** [comment]
In the meantime, thanks to JuMP's quick fix, a rudimentary likelihood
maximizer on this toy problem (sans priors or entropy due to the lgamma
problem) is working JuMPGammaPoisson.jl in the JuMP branch of the repo.
Want to check it out?  I don't yet have an intuitive sense of how fast
something like this should be, but on my laptop it can do 5k points in
about 10 minutes.

I think I need to work on my NIPS paper for the rest of the weekend, so I
may have to pick this up again next week.

Yeah, it's very encouraging how responsive the JuMP team was.
